File: src/main/java/raccoonlang/parser/ast/GenericTypesAstNode.java
package raccoonlang.parser.ast;

import raccoonlang.parser.Parser;
import raccoonlang.tokenizer.Token;
import raccoonlang.tokenizer.TokenType;

import java.util.ArrayList;
import java.util.List;

public class GenericTypesAstNode {

    public List<Token> genericTypeNames = new ArrayList<>();

    public static GenericTypesAstNode tryParse(Parser parser) {
        if(parser.peek().getTokenType() != TokenType.OP_LT) return null;
        parser.take();

        GenericTypesAstNode genericTypesAstNode = new GenericTypesAstNode();

        while (parser.peek().getTokenType() == TokenType.COMMA){
            parser.take();
            genericTypesAstNode.genericTypeNames.add(parser.take(TokenType.IDENTIFIER));
        }

        parser.take(TokenType.OP_GT);
        return genericTypesAstNode;
    }

    @Override
    public String toString() {
        return "GenericTypesAstNode{" +
                "genericTypeNames=" + genericTypeNames.stream().map(x -> x.getText()).reduce((a, b) -> a + ", " + b) +
                '}';
    }
}

File: Parsing/AST/GenericTypesAstNode.cs
namespace Raccoonlang.Parsing.AST;

using Tokenizing;

public class GenericTypesAstNode
{
    public List<Token> GenericTypeNames { get; set; } = new List<Token>();

    public static GenericTypesAstNode? TryParse(Parser parser)
    {
        if (parser.Peek().Type != TokenType.OpLt) return null;
        parser.Take(TokenType.OpLt);

        GenericTypesAstNode node = new GenericTypesAstNode();

        node.GenericTypeNames.Add(parser.Take(TokenType.Identifier));
        while (parser.Peek().Type == TokenType.Comma) {
            parser.Take(TokenType.Comma);
            node.GenericTypeNames.Add(parser.Take(TokenType.Identifier));
        }

        parser.Take(TokenType.OpGt);
        return node;
    }

    public override string ToString()
    {
        return "GenericTypesAstNode{\nGenericTypeNames="+GenericTypeNames.ToArray().ToString()+"\n}";
    }
}